File: 2038-nearest-exit-from-entrance-in-maze/2038-nearest-exit-from-entrance-in-maze.cpp
class Solution {
    struct Pos{
        int x, y, dist; 
    };
    vector<pair<int,int>> dir = {{0,1},{0,-1},{1,0},{-1,0}};
public:
    int nearestExit(vector<vector<char>>& maze, vector<int>& entrance) {
        queue<Pos> q; 
        q.push({entrance[0],entrance[1],0}); 
        vector<vector<bool>> visited(maze.size(), vector<bool>(maze[0].size(),false)); 
        visited[entrance[0]][entrance[1]] = true; 
        while(!q.empty()){
            int size = q.size(); 
            for(int i = 0; i < size; i++){
                Pos first = q.front();
                q.pop(); 
                int x = first.x, y = first.y, dist = first.dist;    

                //cout << x << ' ' << y << ' ' << dist << endl; 

                if(x == 0 || x == maze.size()-1 || y == 0 || y == maze[0].size()-1){
                    
                    if(x != entrance[0] || y != entrance[1]) return dist; 
                }

                for(pair<int,int>& p : dir){
                    int nX = x + p.first; 
                    int nY = y + p.second;  

                    if(nX < 0 || nY < 0 || nX >= maze.size() || nY >= maze[0].size() || maze[nX][nY] != '.' || visited[nX][nY]) continue; 

                    visited[nX][nY] = true; 
                    q.push({nX,nY,dist+1}); 

                }

            }
        }
        return -1; 
    }
};

File: 2038-nearest-exit-from-entrance-in-maze/2038-nearest-exit-from-entrance-in-maze.java
class Solution {
    int[][] dir = {{0,1},{0,-1},{1,0},{-1,0}}; 
    public int nearestExit(char[][] maze, int[] entrance) {
        Queue<int[]> q = new LinkedList<>(); 
        q.add(new int[]{entrance[0],entrance[1],0}); 
        maze[entrance[0]][entrance[1]] = '+'; 

        while(!q.isEmpty()){
            int size = q.size(); 
            for(int i = 0; i < size; i++){
                int[] front = q.poll(); 

                int x = front[0], y = front[1], dist = front[2]; 
                
                for(int[] d : dir){
                    int nX = x + d[0];
                    int nY = y + d[1]; 

                    if(nX < 0 || nY < 0 || nX >= maze.length || nY >= maze[0].length || maze[nX][nY] == '+') continue; 

                    if(nX == 0 || nX == maze.length-1 || nY == 0 || nY == maze[0].length-1) return dist + 1; 

                    maze[nX][nY] = '+'; 
                    q.add(new int[]{nX,nY,dist+1}); 
                }
            }
        }

        return -1; 
    }
}